packages/stylist-macros/src/tokentree/output/mod.rs

```diff
@@ -0,0 +1,44 @@
+use proc_macro2::TokenStream;
+
+mod sheet;
+pub use sheet::OutputSheet;
+mod rule;
+pub use rule::OutputAtRule;
+mod block;
+pub use block::OutputQualifiedRule;
+mod selector;
+pub use selector::OutputQualifier;
+mod scope_content;
+pub use scope_content::OutputScopeContent;
+mod rule_content;
+pub use rule_content::OutputRuleContent;
+mod style_attr;
+pub use style_attr::OutputAttribute;
+mod str_frag;
+pub use str_frag::{fragment_coalesce, fragment_spacing, OutputFragment};
+
+/// Reify a structure into an expression of a specific type.
+pub(crate) trait Reify {
```

Comment:
I can see why you'd want to use `Lazy` by default in the `css!` and `global_style!` macros, but overall, I think it's also overcomplicating things.

I'd be in favor of mentioning that in the documentation and letting the user wrap everything inside the lazy where appropriate:
- even if an expression used, it might be constant (pulling in some config variables, etc...)/could be mentioned in a static constant
- it's easy enough to write something like

```
    static SHEET_REF: Lazy<Sheet> = Lazy::new(|| {
        sheet! {
            &:has(> img) {
                backgroundColor: black;
            }
        }
    });
```
- it actually overcomplicates the behaviour of the macro
- you're writing Rust. If you want performance there is some thought to be given. An macro expanding to an expression, being used in an expression context, expanding to something behind a static Lazy is not my first thought of how it behaves.

Reply:
I don't think it's overcomplicating things.

One fundamental point of macros is to provide optimisation by selecting the most optimised implementation without user intervention.

Example: https://doc.rust-lang.org/src/alloc/macros.rs.html#43

In addition, I don't think simplification need to be made at a cost of user experience.

Why do we need to tell user to optimise themselves if this optimisation can be made by procedural macros automatically?

> Instead of doing
> 
> ```rust
> html! {<div class=css!{ color: red; } />}
> ```
> 
> you should be doing
> 
> ```rust 
> static SHEET: Lazy<Sheet> = Lazy::new(|| sheet! { color: red; });
> html! {<div class=Style::new(SHEET.clone()) />}
> ```
> 
> if you want the best performance.

And imagine you have to do that for every component if you want great performance.
(I cannot imagine anyone would be bothered to do this for every stylesheet if they are maintaining a sizable website.)

> you're writing Rust. If you want performance there is some thought to be given. An macro expanding to an expression, being used in an expression context, expanding to something behind a static Lazy is not my first thought of how it behaves.

How would it behave differently to an end user comparing to a not Lazy'ed variant other than it being slower? And I don't think most users would want to know what code a macro expands to.

In addition, this also helps to mitigate inefficiency in generated code (excessive `StringFragment` when not needed, etc.), in which I will address in the future but haven't had time to do so.

When creating a `Sheet`, there're a lot of small allocations, which is not efficient to do (even worse when people are using `wee_alloc`.). Hence it's somewhat critical to have a `Lazy` in the macro.

```
Benchmark	Result
Parse Simple (1,000,000 iterations):	70ms
Macro Simple (1,000,000 iterations):	5ms
Parse Complex (100,000 iterations):	56ms
Macro Complex (100,000 iterations):	1ms
```

Benchmark when `Lazy` is removed.
```
Benchmark	Result
Parse Simple (1,000,000 iterations):	73ms
Macro Simple (1,000,000 iterations):	79ms (Slower than runtime cache?)
Parse Complex (100,000 iterations):	56ms
Macro Complex (100,000 iterations):	44ms
```

Reply:
Those benchmarks are very persuasive :) I'll have to probably inspect the expressions inside `InjectedExpression`s also, since e.g. `${"2em"}` (a literal fragment being injected) should trigger the optimization as well.